cpp/src/strings/datetime.cu

```diff
@@ -0,0 +1,733 @@
+        if( parse_into_parts(d_str,timeparts) )
+            return 0; // unexpected parse case
+        //
+        return timestamp_from_parts(timeparts,units);
+    }
+};
+
+// convert cudf type to timestamp units
+struct dispatch_timestamp_to_units_fn
+{
+    template <typename T>
+    timestamp_units operator()()
+    {
+        CUDF_FAIL("Invalid type for timestamp conversion.");
+    }
+};
+
+template<>
+timestamp_units dispatch_timestamp_to_units_fn::operator()<cudf::timestamp_D>() { return timestamp_units::days; }
+template<>
+timestamp_units dispatch_timestamp_to_units_fn::operator()<cudf::timestamp_s>() { return timestamp_units::seconds; }
+template<>
+timestamp_units dispatch_timestamp_to_units_fn::operator()<cudf::timestamp_ms>() { return timestamp_units::ms; }
+template<>
+timestamp_units dispatch_timestamp_to_units_fn::operator()<cudf::timestamp_us>() { return timestamp_units::us; }
+template<>
+timestamp_units dispatch_timestamp_to_units_fn::operator()<cudf::timestamp_ns>() { return timestamp_units::ns; }
+
+} // namespace
+
+//
+std::unique_ptr<cudf::column> to_timestamps( strings_column_view strings,
+                                             data_type timestamp_type,
+                                             std::string format,
+                                             rmm::mr::device_memory_resource* mr,
+                                             cudaStream_t stream )
+{
+    size_type strings_count = strings.size();
+    if( strings_count==0 )
+        return make_timestamp_column( timestamp_type, 0 );
+
+    CUDF_EXPECTS( !format.empty(), "Format parameter must not be empty.");
+    timestamp_units units = cudf::experimental::type_dispatcher( timestamp_type, dispatch_timestamp_to_units_fn() );
+
+    format_compiler compiler(format.c_str(),units);
+    format_program* d_prog = compiler.compile_to_device();
+
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = column_device_view::create(strings.parent(), stream);
+    auto d_column = *strings_column;
+
+    // copy null mask
+    rmm::device_buffer null_mask;
+    cudf::size_type null_count = d_column.null_count();
+    if( d_column.nullable() )
+        null_mask = rmm::device_buffer( d_column.null_mask(),
+                                        bitmask_allocation_size_bytes(strings_count),
+                                        stream, mr);
+    // create output column
+    auto results = std::make_unique<cudf::column>( timestamp_type, strings_count,
+        rmm::device_buffer(strings_count * size_of(timestamp_type), stream, mr),
+        null_mask, null_count);
+    auto results_view = results->mutable_view();
+    auto d_results = results_view.data<int64_t>();
+    // set the values
+    thrust::transform( execpol->on(stream),
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(strings_count),
+        d_results, parse_datetime{d_column,d_prog,units} );
```

Comment:
Is `parse_datetime`expected to always parse timestamps in second (or smaller) resolutions? If it can parse into unix-days resolution, we’ll need to adjust this buffer to be 32-bit.

Reply:
Ok, I see. I need to use the type-dispatcher to resolve the `d_results` type correctly based on the timestamp type requested.